rosco-manifests/src/test/java/com/netflix/spinnaker/rosco/manifests/cloudfoundry/CloudFoundryBakeManifestServiceTest.java

```diff
@@ -0,0 +1,136 @@
+
+    when(artifactDownloader.downloadArtifact(any())).thenReturn(manifestTemplate).thenReturn(vars);
+
+    Artifact artifact = cloudFoundryBakeManifestService.bake(request);
+    String resolvedManifest =
+        new String(Base64.getDecoder().decode(artifact.getReference()), "UTF-8");
+    String expectedManifest = "{\"app\":\"bobservice1\",\"version\":\"v999\"}";
+    assertThat(resolvedManifest).asString().isEqualTo(expectedManifest);
+  }
+
+  @Test
+  public void shouldResolveManifestWithMapVars() throws IOException {
+    String manifestString = "{\"app\":\"((app.name))\",\"version\":\"((someversion))\"}";
+    String varsString = "{\"app\":{\"name\":\"bobservice1\"},\"someversion\":\"v999\"}";
+    InputStream manifestTemplate = new ByteArrayInputStream(manifestString.getBytes());
+    InputStream vars = new ByteArrayInputStream(varsString.getBytes());
+
+    CloudFoundryBakeManifestRequest request = new CloudFoundryBakeManifestRequest();
+    request.setManifestTemplate(
+        Artifact.builder()
+            .reference(Base64.getEncoder().encodeToString(manifestString.getBytes()))
+            .build());
+    request.setVarsArtifacts(ImmutableList.of(Artifact.builder().build()));
+
+    when(artifactDownloader.downloadArtifact(any())).thenReturn(manifestTemplate).thenReturn(vars);
+
+    Artifact artifact = cloudFoundryBakeManifestService.bake(request);
+    String resolvedManifest =
+        new String(Base64.getDecoder().decode(artifact.getReference()), "UTF-8");
+    String expectedManifest = "{\"app\":\"bobservice1\",\"version\":\"v999\"}";
+    assertThat(resolvedManifest).asString().isEqualTo(expectedManifest);
+  }
+
+  @Test
+  public void shouldResolveManifestWithListVars() throws IOException {
+    String manifestString = "{\"app\":\"((app[0]))\",\"version\":\"((someversion))\"}";
+    String varsString = "{\"app\":[\"bobservice1\"],\"someversion\":\"v999\"}";
+    InputStream manifestTemplate = new ByteArrayInputStream(manifestString.getBytes());
+    InputStream vars = new ByteArrayInputStream(varsString.getBytes());
+
+    CloudFoundryBakeManifestRequest request = new CloudFoundryBakeManifestRequest();
+    request.setManifestTemplate(
+        Artifact.builder()
+            .reference(Base64.getEncoder().encodeToString(manifestString.getBytes()))
+            .build());
+    request.setVarsArtifacts(ImmutableList.of(Artifact.builder().build()));
+
+    when(artifactDownloader.downloadArtifact(any())).thenReturn(manifestTemplate).thenReturn(vars);
+
+    Artifact artifact = cloudFoundryBakeManifestService.bake(request);
+    String resolvedManifest =
+        new String(Base64.getDecoder().decode(artifact.getReference()), "UTF-8");
+    String expectedManifest = "{\"app\":\"bobservice1\",\"version\":\"v999\"}";
+    assertThat(resolvedManifest).asString().isEqualTo(expectedManifest);
+  }
+
+  @Test
+  public void shouldResolveManifestWithComplexVars() throws IOException {
```

Comment:
Can you add a negative test where the overrides cannot be mapped to the template? I'm not sure what should be expected in the code, but I think end users should see a human readable error in the pipeline execution UI, so that they can fix the mapping without trying to guess what failed.

Reply:
Currently, that feature doesn't exist. If we provided a flag like "failOnUnknownKeys" then we could accomplish this and fail the stage for unknown keys.

Reply:
Since we're doing here a full templating implementation, as opposed to kustomize of helm, where the replacement logic lives in those external tools, I'm trying to step back and see how a user will interact with this feature.
What should happen in this example?:
Manifest:
```
{
  "app": "((appName))",
  "version": "((someversion))"
}
```
Overrides:
```
otherAppName: test
```
I see that not all variables needed by the manifest were supplied by overrides. I'd prefer that spinnaker tells me that the baking failed because of wrong mappings, than trying to send to CF an invalid manifest that could produce unexpected results.
Another overrides example for the same manifest:
```
appName: ""test""
version: 1
```
As I read the code, it will produce invalid json:
```
{
  "app": ""test"",
  "version": "1"
}
```

Reply:
Both the input manifest template and the variables files are required to be yaml files. 
template.yaml
---
`application: ((someName))`

variables.yaml
---
`someName: app1`

This would resolve to:
resolved.yaml
---
`application: app1`

I'm agreeing with you that we should add a feature that would allow the user to select if the stage fails for unknown keys or not.

Reply:
The current logic will try to replace all variables it is aware of so that will need tweaking in order to add a flag for failing on unknown keys. It doesn't currently care if they exist or not.